### PR TITLE
Branch&bound, fixes #1

### LINKED ISSUE #1: add support for integer variables
Adding support for MILP (integer variables) should be quite doable and very useful.

The original author (@ztlpn) already made an implementation using branch & bound, but they only put in in the exemples folder instead of integrating it to the library itself: https://github.com/Specy/microlp/blob/4dc2acc643689730a8451ba9bb303c6437bb16f3/examples/tsp.rs

Integer problems are very common and a tiny implementation like the one in minilp would be very welcome for embedded use cases, such as the wasm.

### PR DESCRIPTION
**Pull request by Specy**
Adds branch and bound to implement binary and integer variables, fixes #1

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -654,4 +714,64 @@ mod tests {
...
+        let entries = vars
+            .iter()
+            .map(|v| (*v, weights[v.0] as f64))
+            .collect::<Vec<_>>();
+        problem.add_constraint(&entries, ComparisonOp::Le, capacity as f64);
+        let sol = problem.solve().unwrap();
+        let values = vars.iter().map(|v| sol[*v]).collect::<Vec<_>>();
+        assert_eq!(
+            values,
+            vec![0.0, 0.0, 0.999999999999999, 0.0, 1.0, 1.0, 1.0, 1.0]
```
Comment (lovasoa) [inline src/lib.rs:738]: are you sure we want the assert with 0.999999999999999 ?

File: src/lib.rs [hunk 2]
```diff
@@ -230,17 +239,50 @@ impl Problem {
...
 
-    /// Add a new variable to the problem.
+    /// Add a new real variable to the problem.
     ///
     /// `obj_coeff` is a coefficient of the term in the objective function corresponding to this
     /// variable, `min` and `max` are the minimum and maximum (inclusive) bounds of this
     /// variable. If one of the bounds is absent, use `f64::NEG_INFINITY` for minimum and
     /// `f64::INFINITY` for maximum.
-    pub fn add_var(&mut self, obj_coeff: f64, (min, max): (f64, f64)) -> Variable {
+    pub fn add_real_var(&mut self, obj_coeff: f64, (min, max): (f64, f64)) -> Variable {
```
Comment (lovasoa) [inline src/lib.rs:253]: I'm not sure this is worth introducing a breaking API change. Can we just keep the existing add_var (just updating the documentation), and add the new functions ? This way we don't need to break the existing minilp api.

File: src/lib.rs [hunk 3]
```diff
@@ -230,17 +239,50 @@ impl Problem {
...
+        let var = self.add_var(obj_coeff, (min, max));
+        self.var_domains.push(VarDomain::Real);
```
Comment (lovasoa) [inline src/lib.rs:255]: this feels dangerous. Shouldn't we let `add_var` push to self.var_domains, so that we never end up in an inconsistent state ?

**Review by lovasoa — CHANGES_REQUESTED** [review]
Really cool :sunglasses: 

I suggested a few changes, feel free to take what you want from them.

File: src/lib.rs (hunk 1 above)
Comment (Specy) [reply src/lib.rs:738]: This was more of a lazy way to just make the test pass. 
But regarding this, once a solution is found, should we round the number to it's integer value or leave it with rounding errors?

Rounding might cause accuracy issues if the solution is reused to add more constraints as we'd have to replace them in the solver state itself.

Or maybe a separate method used to get the variable values which inside clamps only integer variables?

Or nothing at all and we just document the precision issues?

Comment (lovasoa) [reply src/lib.rs:738]: In this case all the variables are integers, I would just cast `values` to a Vec<u8> before the assert.

You could also add a `sol.integer_value(variable)` method.

**Comment by lovasoa** [comment]
also, you may want to add a test with both integer and real variables.

**Comment by Specy** [comment]
> also, you may want to add a test with both integer and real variables.

I wanted to add one but i wanted to use a "known" one or some that already have a solution, do you have one?

**Comment by Specy** [comment]
Ok i added a simple MILP test which is currently failing as it did not reach the optimal value, i'll investigate tomorrow

**Comment by Specy** [comment]
Ok pretty weird, using very large values causes unexpected results. the problem was i64::MAX causing issues, i've clamped integer values to i32 for now and it seems to work properly

the problem seems to happen also when using high f64 values like f64::MAX, but it is unrelated to the integer solution

**Comment by Specy** [comment]
I'm gonna merge this as that bug in #3 was already present so it's not a regression, i'm not sure how to tackle that for now so at least let's support integer variables for now.

I added a failing test to track the bug for the future

File: src/solver.rs [hunk 4]
```diff
@@ -1383,6 +1491,39 @@ fn into_resized(vec: CsVec, len: usize) -> CsVec {
...
 
+fn choose_branch_var(cur_solution: &Solution, domain: &[VarDomain]) -> Option<Variable> {
+    let mut max_divergence = 0.0;
+    let mut max_var = None;
+    for (var, &val) in cur_solution {
+        if domain[var.0] != VarDomain::Integer {
+            continue;
+        }
+        let divergence = f64::abs(val - val.round());
+        if divergence > 1e-5 && divergence > max_divergence {
```
Comment (lovasoa) [inline src/solver.rs:1502]: @Specy I think you should use the EPS constant here, instead of hardcoding 1e-5.

**Pull request merged**